src/cript/api/api.py

```diff
@@ -513,6 +521,11 @@ def save(self, project: Project) -> None:
         None
             Just sends a `POST` or `Patch` request to the API
         """
+
+        # Ensure that all file nodes have uploaded there payload before actual save.
+        for file_node in project.find_children({"node": ["File"]}):
+            file_node.ensure_uploaded(api=self)
```

Comment:
I really dislike this part of the design because it seems confusing to me, but I can accept it for a temporary solution and later we think of something better

Reply:
If you have a better one, let me know.

(One thing, I can think of, that we integrate the `ensure_upload` call into the JSON serialization.
This requires no extra searching of file nodes, they just get uploaded while converted to JSON.
However, this ALWAYS happens BEFORE DB_Schema validation, since we validate the JSON. So we can't perform validation before upload. I don't think that would be great.)

Reply:
oof you are right!

The serialization is a good idea because it makes logical sense, but then we get into the issue that you pointed out earlier that we might upload a ton of files during debugging.

okay I think this should work for now. I think API class needs refactoring anyways because we outgrew our original design, and we can tackle it all there.

Reply:
One thought may be to rewrite the serialization, to conditionally disable the upload there.
(And only activate the upload when the JSON is generated for `save`).

I could implement that.
If you prefer that, I can do that.
Probably takes only a few hours tops.